torch_xla/_patched_functions.py

```diff
@@ -50,5 +50,19 @@ def clip_grad_norm_(parameters: _tensor_or_tensors,
   return total_norm
 
 
+# Overriding the torch's _str method since it performs tensor operations like
+# slice, indexing, etc to make the tensor presentable. These operations
+# on xla tensor results in compilations. Hence, to avoid compilations,
+# copying the tensor to cpu before printing
```

Comment:
Mostly just for my understanding - don't we need to compile and run the graph anyway to print the data in a tensor? Or is the argument just that yes, we need to compile the graph whenever we print, but the extra ops used in the printing process cause noticeable overhead?

Reply:
Yes, the argument is that the extra ops used in the printing process causes an overhead which can be avoided if we just move the tensor to cpu before printing

Reply:
Hmm, doesn't this also cause the printed tensor to show up as being on cpu? (We don't print `device=xla` anymore?)

One other thing I'm wondering about though is composability with other pytorch extensions like functorch, which also defines its own overridden print function. If two pytorch extensions each have their own way of printing a tensor and you import both of them, then you're kind of out of luck.

A few options I can think of are:
(1) Don't override here, and instead update pytorch core's `_tensor_str._str()` to special case xla tensors (and probably other lazy tensors too)
(2) If we do override here, we probably want to make sure it prints the correct device without too many hacks, and also think about composability with functorch (although maybe we can worry about that part later).

(1) seems easier to me. @wconstab what do you think of special casing xla / lazy tensor printing in core to avoid the extra compilations?

Reply:
The print inefficiency has been at the back of my mind for a while now, and I always assumed at some point I'd override print on a python tensor class for lazy tensor- but doing an update to _str() sounds good too, and I hadn't considered the tradeoffs.  I don't think _str is a nativefunction, so can you point me to where it would be overridden in core?

Reply:
yep! That all lives around here: https://github.com/pytorch/pytorch/blob/b16b4448283bcd149829ca833201761c1e1920f3/torch/_tensor_str.py#L428 (it's not a native_function - it's fully implemented in python)

Reply:
Currently in the tensor print, the device is only printed if its not CPU/CUDA. Hence, with this change it would appear as default (since the device won't be printed). 
Agreed, with other extensions (such as functorch) it may become an issue since both will end up overriding the _str method. Hence, I agree the change can be pushed to core PT.
@wconstab do you plan to make the change in core or do you want me to create a PR on that repo?

Reply:
If you're willing to put together a PR that would be awesome.  I think we'd just want a similar case added for the 'lazy' key as the 'xla' key and it would behave the same (call .to cpu before printing).

Reply:
Will put up a PR for handling `xla` tensors. I am not aware of `lazy` key, can you point me to the documentation of these tensors?

Reply:
https://docs.google.com/document/d/1xH5CHuxwSnm-7CWpxHm9FFxsm7Qu-0akAc3o_eX_7uU/edit?usp=sharing

^^ docs on the lazy tensor design

You don't have to handle the lazy part, I can add that subsequently; I was thinking that if you have logic that looks like is_xla()?<copy to cpu>,  you could just include is_lazy() in that logic.  There is already an 'is_lazy' method on tensor. But if it's more involved than that, I'll take care of it.

Reply:
@wconstab @bdhirsh Here is the PR on pytorch core: https://github.com/pytorch/pytorch/pull/71147
Unfortunately, I was not able to add you guys as reviewers there, so just notifying here

Reply:
@wconstab @bdhirsh Here is the PR on torch-core: https://github.com/pytorch/pytorch/pull/71147/files